### ISSUE TITLE
When using `ApplicationInsights::Rack::TrackRequest`, expose a request's ID

### ISSUE THREAD

**Issue by mattrayner** [Microsoft/ApplicationInsights-Ruby#48]
### The Problem
At @ukparliament, we have a number of components which use ApplicationInsights for logging. To help with analytics and monitoring, we'd like to implement hierarchical request IDs as detailed here:

https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/HierarchicalRequestId.md#example

### Proposed Solution
I propose modifying `ApplicationInsights::Rack::TrackRequest#call` so that id generation on line 47: [track_request.rb#L47](https://github.com/Microsoft/ApplicationInsights-Ruby/blob/develop/lib/application_insights/rack/track_request.rb#L47)

Is moved to the top of the method, and that the id is exposed to the application via the `env` hash.

An example could be something like:
```ruby
      # Track requests and send data to Application Insights asynchronously.
      # @param [Hash] env the rack environment.
      def call(env)
        id = rand(16**32).to_s(16)
        env['ApplicationInsight.Request.Id'] = id
        start = Time.now
        begin
          status, headers, response = @app.call(env)
        rescue Exception => ex
          status = 500
        
        # Clipped for brevity

        request = ::Rack::Request.new env
        start_time = start.iso8601(7)
        duration = format_request_duration(stop - start)
        success = status.to_i < 400
        options = {
          :name => "#{request.request_method} #{request.path}",
          :http_method => request.request_method,
          :url => request.url
        }

        @client.track_request id, start_time, duration, status, success, options

        # Clipped for brevity

      end
```

This would enable something like the following from within the application:

```ruby
uri = URI('http://api.example.com/search/?q=test')

req = Net::HTTP::Get.new(uri)
req['Request-Id'] = "|#{env['ApplicationInsight.Request.Id']}." if env['ApplicationInsight.Request.Id']

res = Net::HTTP.start(uri.hostname, uri.port) {|http|
  http.request(req)
}
```

I'd love to know what your opinion is on this?

**Comment by SergeyKanzhelev** [comment]
Directionally we will be switching to W3C standard https://www.w3.org/community/trace-context/ for http headers. If you need interoperability with components in other languages today - yes, this is a good idea. If all components are Ruby - implementing W3C standard may be a better idea.

**Comment by mattrayner** [comment]
At the moment we're using Request-Id headers, so I'll look at implementing the above and opening a PR - in the meantime, I've passed the W3C documentation you mentioned to one of our teams.

**Comment by SergeyKanzhelev** [comment]
Great! Let me know if you need help with PR.

**Comment by langsamu** [comment]
The major use case for this would interoperability with components in other languages: The Ruby application would be calling an ASP.NET Web API application hosted on an Azure App Service, which uses the Application Insights .NET SDK.
@SergeyKanzhelev, what would you recommend in this case?

**Comment by SergeyKanzhelev** [comment]
For now ruby would need to support `Request-Id` header. W3C standard support will most probably be released by the end of this calendar year.

**Comment by mattrayner** [comment]
Off the back of this, I’ll work on a pull request tonight/tomorrow and send
it over

**Comment by mattrayner** [comment]
Fixed by [#54]
